drv/lpc55-update-server/src/main.rs

```diff
@@ -279,31 +277,133 @@ fn validate_header_block(
     Ok(())
 }
 
+/// Performs an erase-write sequence to a single page within a given target
+/// image.
 fn do_block_write(
     img: UpdateTarget,
     block_num: usize,
     flash_page: &mut [u8; BLOCK_SIZE_BYTES],
 ) -> Result<(), UpdateError> {
-    let result = unsafe {
-        // The write_to_flash API takes raw pointers due to TrustZone
-        // ABI requirements which makes this function unsafe.
-        tz_table!().write_to_flash(
-            img,
-            block_num as u32,
-            flash_page.as_mut_ptr(),
-        )
+    let page_num = block_num as u32; // TODO why not use the same type
+
+    let flash = unsafe { &*lpc55_pac::FLASH::ptr() };
+    let mut flash = drv_lpc55_flash::Flash::new(flash);
+
+    // Can only update opposite image
+    if same_image(img) {
+        return Err(UpdateError::RunningImage);
+    }
+
+    let write_addr = match target_addr(img, page_num) {
+        Some(addr) => addr,
+        None => return Err(UpdateError::OutOfBounds),
     };
 
-    match result {
-        HypoStatus::Success => Ok(()),
-        HypoStatus::OutOfBounds => Err(UpdateError::OutOfBounds),
-        HypoStatus::RunningImage => Err(UpdateError::RunningImage),
-        // Should probably encode the LPC55 flash status into the update
-        // error for good measure but that takes effort...
-        HypoStatus::FlashError(_) => Err(UpdateError::FlashError),
+    // Step one: erase the page. 
+
+    // write_addr is a byte address; page_num is a page index; but the actual
+    // erase machinery operates in terms of flash words. A flash word is 16
+    // bytes in length. So, we need to convert.
+    //
+    // Note that the range used here is _inclusive._
+    //
+    // We wind up needing this in u32 form a lot, so let's cast once and reuse.
+    const WORDSZ: u32 = BYTES_PER_FLASH_WORD as u32;
```

Comment:
In the interest of paranoia, it may be worthwhile to have a
```rust
static_assertions::const_assert_eq!(BLOCK_SIZE_BYTES % BYTES_PER_FLASH_WORD, 0)
```
somewhere near here.